backport: T4515: T4219: policy local-route6 and inbound-interface support

hensur opened the pull request:
Sorry for the inconvenience, I hadn't have much time the past few days.
This should finally fix the issues from my backport attempt.
I think I messed up a rebase or merge that caused this additional check to appear again.

https://github.com/vyos/vyos-1x/pull/1235

<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://phabricator.vyos.net/Txxxx

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->

## Proposed changes
<!--- Describe your changes in detail -->

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [ ] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [ ] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [ ] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

hensur committed:
policy: T4151: remove left over fwmk check

hensur committed:
Revert "Revert "backport: T4515: T4219: policy local-route6 and inbound-interface support""

This reverts commit 45a2a7d0adc7e9d27d6c7aee1ccbd9b64a1437ad.

hensur commented:
```
vyos@vyos:~$ /usr/libexec/vyos/tests/smoke/cli/test_policy.py
test_access_list (__main__.TestPolicy) ... ok
test_access_list6 (__main__.TestPolicy) ... ok
test_as_path_list (__main__.TestPolicy) ... ok
test_community_list (__main__.TestPolicy) ... ok
test_delete_ipv4_ipv6_table_id (__main__.TestPolicy) ... ok
test_destination_ipv6_table_id (__main__.TestPolicy) ... ok
test_destination_table_id (__main__.TestPolicy) ... ok
test_extended_community_list (__main__.TestPolicy) ... ok
test_fwmark_ipv6_table_id (__main__.TestPolicy) ... ok
test_fwmark_sources_destination_ipv6_table_id (__main__.TestPolicy) ... ok
test_fwmark_sources_destination_table_id (__main__.TestPolicy) ... ok
test_fwmark_sources_ipv6_table_id (__main__.TestPolicy) ... ok
test_fwmark_sources_table_id (__main__.TestPolicy) ... ok
test_fwmark_table_id (__main__.TestPolicy) ... ok
test_iif_sources_ipv6_table_id (__main__.TestPolicy) ... ok
test_iif_sources_table_id (__main__.TestPolicy) ... ok
test_ipv6_table_id (__main__.TestPolicy) ... ok
test_large_community_list (__main__.TestPolicy) ... ok
test_multiple_commit_ipv4_table_id (__main__.TestPolicy) ... ok
test_prefix_list (__main__.TestPolicy) ... ok
test_prefix_list6 (__main__.TestPolicy) ... ok
test_table_id (__main__.TestPolicy) ... ok

----------------------------------------------------------------------
Ran 22 tests in 51.417s

OK
```

sever-sever commented:
Doesn't work properly
Initial configuration:
```
set policy local-route rule 110 destination '203.0.113.0/24'
set policy local-route rule 110 fwmark '123'
set policy local-route rule 110 set table '170'
set policy local-route rule 110 source '192.0.5.22'
```
IP rule:
```
vyos@testrouter# sudo ip rule show
0:	from all lookup local
110:	from 192.0.5.22 to 203.0.113.0/24 fwmark 0x7b lookup 170
32766:	from all lookup main
32767:	from all lookup default
```

Delete source and we still see source address 192.0.5.22:
```
vyos@testrouter# delete policy local-route rule 110 source 
[edit]
vyos@testrouter# commit
[edit]
vyos@testrouter# sudo ip rule show
0:	from all lookup local
110:	from 192.0.5.22 to 203.0.113.0/24 fwmark 0x7b lookup 170
32766:	from all lookup main
32767:	from all lookup default
[edit]
vyos@testrouter# 
```

What we expect after deleting the source:
```
vyos@testrouter# sudo ip rule show
0:	from all lookup local
110:	from all to 203.0.113.0/24 fwmark 0x7b lookup 170
32766:	from all lookup main
32767:	from all lookup default
```

sever-sever commented:
Doesn't work properly
Initial configuration:
```
set policy local-route rule 110 destination '203.0.113.0/24'
set policy local-route rule 110 fwmark '123'
set policy local-route rule 110 set table '170'
set policy local-route rule 110 source '192.0.5.22'
```
IP rule:
```
vyos@testrouter# sudo ip rule show
0:	from all lookup local
110:	from 192.0.5.22 to 203.0.113.0/24 fwmark 0x7b lookup 170
32766:	from all lookup main
32767:	from all lookup default
```

Delete source and we still see source address 192.0.5.22:
```
vyos@testrouter# delete policy local-route rule 110 source 
[edit]
vyos@testrouter# commit
[edit]
vyos@testrouter# sudo ip rule show
0:	from all lookup local
110:	from 192.0.5.22 to 203.0.113.0/24 fwmark 0x7b lookup 170
32766:	from all lookup main
32767:	from all lookup default
[edit]
vyos@testrouter# 
```

What we expect after deleting the source:
```
vyos@testrouter# sudo ip rule show
0:	from all lookup local
110:	from all to 203.0.113.0/24 fwmark 0x7b lookup 170
32766:	from all lookup main
32767:	from all lookup default
```

hensur committed:
correctly set fwmk from current config on removals, fix smoketest

hensur commented:
@sever-sever Sorry about that. Good catch :) On removal of other options, fwmark should be set to the current value, however, this was a string type, while a list is returned if fwmark itself gets deleted. This happens for inbound-interface as well.
Untyped python is still confusing me sometimes.

Both are fixed by 45734d25f6b4f930fbc572be7ab247a377e179bf
I also added a smoketest and fixed two tests, which were inserted wrongly after my rebase (I think).

sever-sever commented:
@hensur it still incorrect behaviour
```
set policy local-route rule 110 destination '203.0.113.0/24'
set policy local-route rule 110 set table '100'
commit
set policy local-route rule 110 set table '101'
commit

```
Expected table 101 only:
```
110:	from all to 203.0.113.0/24 lookup 100
110:	from all to 203.0.113.0/24 lookup 101
```

Could you do more test? 
Thanks

hensur committed:
respect table changes for remove_rule

hensur commented:
Oh, well. There never was a check for differences in the table during remove, so I didn't think to add one as well.

I just implemented one, along with a matching smoketest.

sever-sever reviewed (approved):
Thanks
All tests looks good

andamasov commented:
@dmbaturin @zdc can you check as well?

initramfs commented on interface-definitions/policy-local-route.xml.in:
```diff
@@ -58,6 +70,115 @@
+                <properties>
+                  <help>Source address or prefix</help>
+                  <valueHelp>
+                    <format>ipv4</format>
```
This should be `ipv6`

hensur replied:
Thanks, I probably missed this while testing because the validation is using ipv6 already.

initramfs commented on interface-definitions/policy-local-route.xml.in:
```diff
@@ -58,6 +70,115 @@
+                    <description>Address to match against</description>
+                  </valueHelp>
+                  <valueHelp>
+                    <format>ipv4net</format>
```
This should be `ipv6net`

initramfs commented:
Tried to use this patch, running into issues with invalid commit application ordering between `inbound-interface` and the creation of the interface. To expand further I have a bonding interface with a VLAN on it similar to this:

```
interfaces {
    bonding bond0 {
        ...
        vif 10 {
            ...
        }
    }
}
```

When a policy local-route or local-route6 uses the bonding VLAN interface (`bond0.10` in the example) as an `inbound-interface` like so:
```
policy {
    local-route {
        rule 1 {
            ...
            inbound-interface bond0.10
        }
    }
}
```
Configuration fails at boot time claiming the interface does not exist. Trying to repair the configuration requires creating the interface, committing, then adding the policy back before the config applies.

**Edit:** Upon further testing the issue occurs both on a pure (non-VLAN) bonding interface as well as an ethernet VLAN interface, suggesting that the incompatibility exists whenever the interface is created at 'runtime'. The same issue is present in 1.4 as well, with the bug report [here](https://phabricator.vyos.net/T4685).

hensur committed:
local-route6: use ipv6 value help for source

hensur committed:
policy: local-route(6): set priority property

Co-authored-by: initramfs <initramfs@initramfs.io>

hensur commented:
@initramfs I've backported the changes from https://github.com/vyos/vyos-1x/pull/1532/files as well. I guess that should fix the issue for 1.3 as well. Thank you!
